src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,202 @@
+
+    public override BoundExpression InstrumentForStatementCondition(BoundForStatement original, BoundExpression rewrittenCondition, SyntheticBoundNodeFactory factory)
+        => InstrumentExpression(base.InstrumentForStatementCondition(original, rewrittenCondition, factory));
+
+    public override BoundStatement InstrumentForStatementConditionalGotoStartOrBreak(BoundForStatement original, BoundStatement branchBack)
+        => InstrumentStatement(base.InstrumentForStatementConditionalGotoStartOrBreak(original, branchBack));
+
+    public override BoundStatement InstrumentForEachStatementConditionalGotoStart(BoundForEachStatement original, BoundStatement branchBack)
+        => InstrumentStatement(base.InstrumentForEachStatementConditionalGotoStart(original, branchBack));
+
+    public override BoundStatement InstrumentGotoStatement(BoundGotoStatement original, BoundStatement rewritten)
+        => InstrumentStatement(base.InstrumentGotoStatement(original, rewritten));
+
+    public override void InterceptCallAndAdjustArguments(
+        ref MethodSymbol method,
+        ref BoundExpression? receiver,
+        ref ImmutableArray<BoundExpression> arguments,
+        ref ImmutableArray<RefKind> argumentRefKindsOpt)
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt);
+
+        if (arguments is [.., { Type: { } lastArgumentType } lastArgument] &&
+            (argumentRefKindsOpt.IsDefault || argumentRefKindsOpt is [.., RefKind.None]) &&
+            lastArgumentType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+        {
+            // The last argument is a CancellationToken. Replace it with the module-level token.
+            // Keep the previous expression so that side-effects are preserved.
+            arguments = [.. arguments[0..^1], _factory.Sequence([lastArgument], _factory.ModuleCancellationToken())];
+        }
+        else if (FindOverloadWithCancellationToken(method) is { } cancellableOverload)
+        {
+            // The method being invoked does not have a CancellationToken as the last parameter, but there is an overload that does.
+            // Invoke the other overload instead and pass in module-level token. 
+            method = cancellableOverload;
+            arguments = [.. arguments, _factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        if (method.MethodKind is not (MethodKind.Ordinary or MethodKind.Constructor or MethodKind.AnonymousFunction or MethodKind.LocalFunction))
```

Comment:
>(MethodKind.Ordinary or MethodKind.Constructor or MethodKind.AnonymousFunction or MethodKind.LocalFunction)) [](http://example.com/codeflow?start=37&length=108)

It would be good to add tests observing presence of this condition.

Reply:
Only kept `MethodKind.Ordinary or MethodKind.Constructor`.